lib/cloud.js

```diff
@@ -1,23 +1,43 @@
-var path = require('path')
-var hyperarchiver = require('hyperarchiver')
+var assert = require('assert')
+var fs = require('fs')
+var memdb = require('memdb')
+var Archiver = require('hypercore-archiver')
+// var ArchiverServer = require('archiver-server') TODO needed?
+var ArchiverApi = require('archiver-api')
 
 module.exports = Hypercloud
 
 function Hypercloud (opts) {
   if (!(this instanceof Hypercloud)) return new Hypercloud(opts)
-  if (!opts) opts = {}
-  var self = this
 
-  opts.dir = opts.dir || path.join(__dirname, 'hypercloud')
-  var hyper = hyperarchiver(opts)
+  assert.ok(opts, 'hypercloud requires options')
+  assert.ok(opts.dir || opts.memdb, 'hyperarchiver requires a dir or memdb option')
 
-  self.archiver = hyper.archiver
-  self.dat = hyper.dat
-  self.api = hyper.api
+  var dirOrDb
+  if (opts.dir) {
+    // ensure the target dir exists
+    dirOrDb = opts.dir
+    try {
+      fs.accessSync(opts.dir, fs.F_OK)
+    } catch (e) {
+      fs.mkdirSync(opts.dir)
+    }
+  } else {
+    // create the memdb
+    dirOrDb = memdb()
+  }
+
+  this.archiver = Archiver(dirOrDb)
```

Comment:
I think this may have changed, but it'd be nice to separate the directory where the database is (or mem) and the storage directory (dir/data currently). Having both of these passed in as one is a bit confusing.